File: leetcode/0170-two-sum-iii-data-structure-design/solution.cpp
#include <unordered_map>
using namespace std;

class TwoSum {
public:
    TwoSum() {}

    void add(int number) {
        counts[number]++;
    }

    bool find(int value) {
        for (auto& [k, cnt] : counts) {
            int complement = value - k;
            if (complement == k) {
                if (cnt > 1) return true;
            } else if (counts.count(complement)) {
                return true;
            }
        }
        return false;
    }

private:
    unordered_map<int, int> counts;
};

/**
 * Your TwoSum object will be instantiated and called as such:
 * TwoSum* obj = new TwoSum();
 * obj->add(number);
 * bool param_2 = obj->find(value);
 */

File: leetcode/0170-two-sum-iii-data-structure-design/solution.py
from collections import defaultdict


class TwoSum:
    def __init__(self):
        self.counts = defaultdict(int)

    def add(self, number: int) -> None:
        self.counts[number] += 1

    def find(self, value: int) -> bool:
        for k in self.counts:
            complement = value - k
            if complement == k:
                if self.counts[k] > 1:
                    return True
            elif complement in self.counts:
                return True
        return False


# Your TwoSum object will be instantiated and called as such:
# obj = TwoSum()
# obj.add(number)
# param_2 = obj.find(value)
